io/src/lib.rs

```diff
@@ -67,6 +80,28 @@ impl<'a, R: Read + ?Sized> Read for Take<'a, R> {
     }
 }
 
+// Impl copied from Rust stdlib.
+impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
+    #[inline]
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        // Don't call into inner reader at all at EOF because it may still block
+        if self.remaining == 0 {
+            return Ok(&[]);
+        }
+
+        let buf = self.reader.fill_buf()?;
+        let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
```

Comment:
I know why it's written like this but could have a comment about it so someone doesn't come up with a "bright" idea of changing this to `cmp::min(buf.len(), self.remaining as usize)`.

Reply:
Alright, you are going to have to help me out here. I've looked at this a couple of separate times now and I can't find the case where the following two lines create a different `cap` value.
```rust
let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
let cap = cmp::min(buf.len(), self.remaining as usize);
```

Reply:
The latter may overflow on 32 bit architecture.

Reply:
Ahh, because `Read::take` accepts a `u64`. Thanks.

Reply:
Added the following comment
```rust
        // Cast length to a u64 instead of casting `remaining` to a `usize`
        // (in case `remaining > u32::MAX` and we are on a 32 bit machine).
        let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
```